core/src/main/java/org/apache/calcite/sql/validate/DelegatingScope.java

```diff
@@ -542,6 +543,208 @@ private int worstKind(Path path) {
     }
   }
 
+
+  /**
+   * Converts a table identifier with an ID column into a fully-qualified identifier.
+   * For example, the dept in "select empno from emp natural join dept" may become
+   * "myschema.dept".
+   *
+   * @param identifier SqlTableIdentifierWithID to qualify
+   * @return A qualified identifier, never null
+   */
+  @Override public SqlTableIdentifierWithIDQualified fullyQualify(
+      SqlTableIdentifierWithID identifier) {
+    // This implementation is based on the fullyQualify for regular identifiers above. We skip
+    // anything that only pertains to columns because we are only handling tables.
+    final SqlTableIdentifierWithID previous = identifier;
+    final SqlNameMatcher nameMatcher = validator.catalogReader.nameMatcher();
+    SqlValidatorNamespace fromNs = null;
+    Path fromPath = null;
+    RelDataType fromRowType = null;
+    final ResolvedImpl resolved = new ResolvedImpl();
+    int size = identifier.getNames().size();
+    int i = size - 1;
+    for (; i > 0; i--) {
```

Comment:
So, we're just going to the fall through case in the original fullyQualify (line 306). Why?

Reply:
This implies that identifier.names.size() always != 1, or we always fall through. I don't believe that identifier.names.size() always != 1, so we must always fallthrough. The section that is omited seems to me, at a high level, to attempt to resolve the identifier in the some different scopes given by findQualifyingTableNames (which, in delegating scope, just calls the parent scope), before going to the default.

So, my guess is that we're allowed to skip this because the merge into table is the highest level scope/doesn't have a parent? I'm still not fully sure about this, am I on the right track?

Reply:
We had a conversation about this, marking as resolved.